flink-clients/src/main/java/org/apache/flink/client/deployment/application/EmbeddedClient.java

```diff
@@ -0,0 +1,154 @@
+public class EmbeddedClient implements JobClient {
+
+	private static final Logger LOG = LoggerFactory.getLogger(EmbeddedClient.class);
+
+	private final JobID jobId;
+
+	private final Configuration configuration;
+
+	private final DispatcherGateway dispatcherGateway;
+
+	private final Time timeout;
+
+	public EmbeddedClient(
+			final JobID jobId,
+			final Configuration configuration,
+			final DispatcherGateway dispatcherGateway) {
+		this.jobId = checkNotNull(jobId);
+		this.configuration = checkNotNull(configuration);
+		this.dispatcherGateway = checkNotNull(dispatcherGateway);
+		this.timeout = Time.milliseconds(configuration.getLong(WebOptions.TIMEOUT));
+	}
+
+	public CompletableFuture<JobID> submitJob(final JobGraph jobGraph) {
+		checkNotNull(jobGraph);
+
+		LOG.info("Submitting Job with JobId={}.", jobGraph.getJobID());
+
+		return dispatcherGateway
+				.getBlobServerPort(timeout)
+				.thenApply(blobServerPort -> new InetSocketAddress(dispatcherGateway.getHostname(), blobServerPort))
+				.thenCompose(blobServerAddress -> {
+
+					try {
+						ClientUtils.extractAndUploadJobGraphFiles(jobGraph, () -> new BlobClient(blobServerAddress, configuration));
+					} catch (FlinkException e) {
+						throw new CompletionException(e);
+					}
+
+					return dispatcherGateway.submitJob(jobGraph, timeout);
+				}).thenApply(ack -> jobGraph.getJobID());
+	}
+
+	@Override
+	public JobID getJobID() {
+		return jobId;
+	}
+
+	@Override
+	public CompletableFuture<JobStatus> getJobStatus() {
+		return dispatcherGateway.requestJobStatus(jobId, timeout);
+	}
+
+	@Override
+	public CompletableFuture<Void> cancel() {
+		return dispatcherGateway
+				.cancelJob(jobId, timeout)
+				.thenApply(ignores -> null);
+	}
+
+	@Override
+	public CompletableFuture<String> stopWithSavepoint(final boolean advanceToEndOfEventTime, @Nullable final String savepointDirectory) {
+		return dispatcherGateway.stopWithSavepoint(jobId, savepointDirectory, advanceToEndOfEventTime, timeout);
+	}
+
+	@Override
+	public CompletableFuture<String> triggerSavepoint(@Nullable final String savepointDirectory) {
+		return dispatcherGateway.triggerSavepoint(jobId, savepointDirectory, false, timeout);
+	}
+
+	@Override
+	public CompletableFuture<Map<String, Object>> getAccumulators(final ClassLoader classLoader) {
+		checkNotNull(classLoader);
+
+		return dispatcherGateway.requestJob(jobId, timeout)
+				.thenApply(ArchivedExecutionGraph::getAccumulatorsSerialized)
+				.thenApply(accumulators -> {
+					try {
+						return AccumulatorHelper.deserializeAndUnwrapAccumulators(accumulators, classLoader);
+					} catch (Exception e) {
+						throw new CompletionException("Cannot deserialize and unwrap accumulators properly.", e);
+					}
+				});
+	}
+
+	@Override
+	public CompletableFuture<JobExecutionResult> getJobExecutionResult(final ClassLoader userClassloader) {
+		checkNotNull(userClassloader);
+
+		return dispatcherGateway
+				.requestJobResult(jobId, RpcUtils.INF_TIMEOUT)
```

Comment:
I know the background that when we later in application mode scenario we have to retry and kill the application on terminate. But I think it is wried for others who will ask why `RpcUtils.INF_TIMEOUT` here.

Also, even we set to `RpcUtils.INF_TIMEOUT` it is actually about 9 months and possibly still we need retry on retiable exception(AskTimeout).

Reply:
You mean that we should go with a solution that retries periodically if the result is not there?

Reply:
@TisonKun given that we are using directly the `dispatcher`, I am not so sure even if the `timeout` plays a role.

Reply:
Every gateway instance is a proxy so I think `timeout` always works.

>You mean that we should go with a solution that retries periodically if the result is not there?

Yes if we meet a retriable exception.

For another perspective, if this method call by user, why not use configured `timeout`; if this method call by framework, you can take a look at `RestClusterClient#requestJobResult` where there is a retry mechanism underline.

Reply:
I think you are right @TisonKun . I will update this but given that for now we are not using this method and we do not want to block a thread if the user does wait for the result by using explicitly the `JobClient`, I change the PR to return a special client which does not allow to fetch the result.

I will integrate the retry logic in a follow-up PR about the `StandaloneJobClusterEntrypoint`.

Reply:
That should make sense.